superset-frontend/packages/superset-ui-chart-controls/src/shared-controls/dndControls.tsx

```diff
@@ -83,8 +89,12 @@ export const dnd_adhoc_filters: SharedControlConfig<'DndFilterSelect'> = {
   default: [],
   description: '',
   mapStateToProps: ({ datasource, form_data }) => ({
-    columns: datasource?.columns.filter(c => c.filterable) || [],
-    savedMetrics: datasource?.metrics || [],
+    columns: datasource?.columns[0]?.hasOwnProperty('filterable')
+      ? (datasource as Dataset)?.columns.filter(c => c.filterable)
+      : datasource?.columns || [],
+    savedMetrics: datasource?.hasOwnProperty('metrics')
```

Comment:
do you think we can dry up these three?
```
datasource?.hasOwnProperty('metrics')
      ? (datasource as Dataset)?.metrics || []
      : DEFAULT_METRICS,
```

Reply:
Done in [`this commit`](https://github.com/apache/superset/pull/19855/commits/a5855462a4ba47258f598734a94c9df385715ea1) 😁